File: .cursor/hooks/state-watcher.py
#!/usr/bin/env python3
"""postToolUse hook: validate workflow-state edits against the schema.

Fires after a tool finishes. When the touched file basename is
`workflow-state.json`, the hook re-reads the file from disk and checks it
against `.cursor/state/workflow-state.schema.json` using only the Python
standard library (no external `jsonschema` dependency). The hook is purely
observational: it emits `additional_context` describing the validation
result and never edits any file.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path, PurePosixPath

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _trace import trace as _trace  # noqa: E402

ROOT = Path(__file__).resolve().parents[2]
SCHEMA_PATH = ROOT / ".cursor" / "state" / "workflow-state.schema.json"

EDIT_TOOLS = {"Write", "Edit", "MultiEdit", "NotebookEdit"}


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def _extract_tool_name(payload: dict) -> str:
    for key in ("tool_name", "toolName", "name"):
        value = payload.get(key)
        if isinstance(value, str) and value:
            return value
    return ""


def _extract_file_path(payload: dict) -> str:
    tool_input = payload.get("tool_input") if isinstance(payload, dict) else None
    if isinstance(tool_input, dict):
        for key in ("file_path", "filePath", "path", "notebook_path", "notebookPath"):
            value = tool_input.get(key)
            if isinstance(value, str) and value.strip():
                return value
    for key in ("file_path", "filePath", "path"):
        value = payload.get(key)
        if isinstance(value, str) and value.strip():
            return value
    return ""


def _validate_against_schema(document: object, schema: dict) -> list[str]:
    errors: list[str] = []
    if not isinstance(document, dict):
        errors.append("state document root must be an object")
        return errors
    required = schema.get("required") if isinstance(schema, dict) else None
    if isinstance(required, list):
        for field in required:
            if field not in document:
                errors.append(f"missing required field: {field}")
    properties = schema.get("properties", {}) if isinstance(schema, dict) else {}
    if not isinstance(properties, dict):
        return errors
    for field, value in document.items():
        spec = properties.get(field)
        if not isinstance(spec, dict):
            continue
        enum = spec.get("enum")
        if isinstance(enum, list) and value not in enum:
            errors.append(f"field {field} must be one of {sorted([str(item) for item in enum])}")
    criteria = document.get("acceptance_criteria")
    if isinstance(criteria, list):
        ac_statuses = {"pending", "passed", "failed", "skipped"}
        for index, item in enumerate(criteria):
            if not isinstance(item, dict):
                errors.append(f"acceptance_criteria[{index}] must be an object")
                continue
            for key in ("id", "criterion", "status"):
                if key not in item:
                    errors.append(f"acceptance_criteria[{index}] missing required field: {key}")
            status = item.get("status")
            if isinstance(status, str) and status not in ac_statuses:
                errors.append(f"acceptance_criteria[{index}].status must be one of {sorted(ac_statuses)}")
    failure = document.get("failure")
    if isinstance(failure, dict):
        retry = failure.get("retry_count")
        if isinstance(retry, int) and not (0 <= retry <= 3):
            errors.append("failure.retry_count must be between 0 and 3")
    return errors


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({"status": "pass", "fail_open": True, "additional_context": "", "message": "State-watcher input was not JSON; skipped."}))
        return 0

    tool_name = _extract_tool_name(payload)
    file_path = _extract_file_path(payload)
    basename = PurePosixPath(file_path.replace("\\", "/")).name if file_path else ""

    if tool_name not in EDIT_TOOLS or basename != "workflow-state.json":
        _trace({"hook": "state-watcher", "event": "postToolUse", "status": "pass", "checked": False, "tool_name": tool_name, "file_basename": basename})
        print(json.dumps({"status": "pass", "fail_open": True, "additional_context": "", "checked": False}))
        return 0

    target = Path(file_path).expanduser()
    if not target.is_absolute():
        target = (ROOT / target).resolve()

    if not target.is_file():
        print(json.dumps({
            "status": "pass",
            "fail_open": True,
            "additional_context": "State-watcher could not read the edited workflow-state file from disk.",
            "checked": False,
            "file_path": file_path,
        }, ensure_ascii=False))
        return 0

    try:
        document = json.loads(target.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        print(json.dumps({
            "status": "pass",
            "fail_open": True,
            "additional_context": f"State-watcher could not parse the workflow-state JSON: {exc}.",
            "checked": False,
            "file_path": file_path,
        }, ensure_ascii=False))
        return 0

    schema: dict = {}
    if SCHEMA_PATH.is_file():
        try:
            schema = json.loads(SCHEMA_PATH.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            schema = {}

    errors = _validate_against_schema(document, schema if isinstance(schema, dict) else {})
    if errors:
        additional_context = (
            "State-watcher detected schema issues in the edited workflow-state document: "
            + "; ".join(errors[:8])
            + ". Use scripts/validate-workflow-state.py for the full validator output."
        )
        status = "warning"
    else:
        additional_context = (
            "State-watcher confirmed the edited workflow-state document still matches "
            ".cursor/state/workflow-state.schema.json."
        )
        status = "pass"

    _trace({
        "hook": "state-watcher",
        "event": "postToolUse",
        "status": status,
        "checked": True,
        "tool_name": tool_name,
        "file_basename": basename,
        "error_count": len(errors),
        "first_errors": errors[:4],
    })
    print(json.dumps({
        "status": status,
        "fail_open": True,
        "checked": True,
        "additional_context": additional_context,
        "file_path": file_path,
        "errors": errors[:16],
    }, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/state-watcher.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';
import { resolveRepoRoot, resolveWorkspaceRoot } from './_repo.ts';
import { trace } from './_trace.ts';
import { extractFilePath, extractToolName } from './_tool_payload.ts';

const currentFile = fileURLToPath(import.meta.url);
const PAYLOAD_ROOT = resolveRepoRoot(currentFile);
const WORKSPACE_ROOT = resolveWorkspaceRoot(currentFile);

const SCHEMA_PATH = path.join(PAYLOAD_ROOT, '.cursor', 'state', 'workflow-state.schema.json');
const DEFAULT_STATE_PATH = path.join(WORKSPACE_ROOT, '.cursor', 'state', 'workflow-state.json');

const EDIT_TOOLS = new Set(['Write', 'Edit', 'MultiEdit', 'NotebookEdit']);

const BRIDGE_WRITE_TOOLS = new Set([
  'state_init',
  'state_set_phase',
  'state_record_failure',
  'state_update_acceptance_criterion',
  'state_history_append',
]);

const MCP_PREFIX = 'mcp__cursor-state-bridge__';

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function resolveBridgeCall(toolName: string, payload: any): [string, string] {
  let inner = '';
  if (toolName.startsWith(MCP_PREFIX)) {
    inner = toolName.slice(MCP_PREFIX.length);
  } else if (toolName === 'tools/call') {
    const toolInput = payload?.tool_input;
    if (toolInput && typeof toolInput === 'object') {
      const candidate = toolInput.name;
      if (typeof candidate === 'string') {
        inner = candidate.startsWith(MCP_PREFIX) ? candidate.slice(MCP_PREFIX.length) : candidate;
      }
    }
  } else if (BRIDGE_WRITE_TOOLS.has(toolName)) {
    inner = toolName;
  }

  if (!BRIDGE_WRITE_TOOLS.has(inner)) {
    return ['', ''];
  }

  let argumentsObj: any = {};
  const toolInput = payload?.tool_input;
  if (toolInput && typeof toolInput === 'object') {
    const nested = toolInput.arguments;
    if (nested && typeof nested === 'object' && !Array.isArray(nested)) {
      argumentsObj = nested;
    } else {
      argumentsObj = toolInput;
    }
  }
  const taskId = typeof argumentsObj?.task_id === 'string' ? argumentsObj.task_id : '';
  if (taskId) {
    return [inner, path.join(WORKSPACE_ROOT, 'docs', 'plans', taskId, 'workflow-state.json')];
  }
  return [inner, DEFAULT_STATE_PATH];
}

function validateAgainstSchema(document: any, schema: any): string[] {
  const errors: string[] = [];
  if (!document || typeof document !== 'object' || Array.isArray(document)) {
    errors.push('state document root must be an object');
    return errors;
  }
  const required = schema?.required;
  if (Array.isArray(required)) {
    for (const field of required) {
      if (!(field in document)) {
        errors.push(`missing required field: ${field}`);
      }
    }
  }
  const properties = schema?.properties;
  if (properties && typeof properties === 'object' && !Array.isArray(properties)) {
    for (const [field, value] of Object.entries(document)) {
      const spec = properties[field];
      if (spec && typeof spec === 'object' && !Array.isArray(spec)) {
        const enumList = spec.enum;
        if (Array.isArray(enumList) && !enumList.includes(value)) {
          errors.push(`field ${field} must be one of ${[...enumList].sort().map(String).join(', ')}`);
        }
      }
    }
  }
  const criteria = document.acceptance_criteria;
  if (Array.isArray(criteria)) {
    const acStatuses = new Set(['pending', 'passed', 'failed', 'skipped']);
    for (let index = 0; index < criteria.length; index++) {
      const item = criteria[index];
      if (!item || typeof item !== 'object' || Array.isArray(item)) {
        errors.push(`acceptance_criteria[${index}] must be an object`);
        continue;
      }
      for (const key of ['id', 'criterion', 'status']) {
        if (!(key in item)) {
          errors.push(`acceptance_criteria[${index}] missing required field: ${key}`);
        }
      }
      const status = item.status;
      if (typeof status === 'string' && !acStatuses.has(status)) {
        errors.push(`acceptance_criteria[${index}].status must be one of ${[...acStatuses].sort().join(', ')}`);
      }
    }
  }
  const failure = document.failure;
  if (failure && typeof failure === 'object' && !Array.isArray(failure)) {
    const retry = failure.retry_count;
    if (typeof retry === 'number' && (retry < 0 || retry > 3)) {
      errors.push('failure.retry_count must be between 0 and 3');
    }
  }
  return errors;
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      additional_context: '',
      message: 'State-watcher input was not JSON; skipped.'
    }));
    return 0;
  }

  const toolName = extractToolName(payload);
  const filePath = extractFilePath(payload);
  const basename = filePath ? path.posix.basename(filePath.replace(/\\/g, '/')) : '';

  const [bridgeTool, bridgeTarget] = resolveBridgeCall(toolName, payload);
  const isDirectEdit = EDIT_TOOLS.has(toolName) && basename === 'workflow-state.json';

  if (!isDirectEdit && !bridgeTool) {
    trace({
      hook: 'state-watcher',
      event: 'postToolUse',
      status: 'pass',
      checked: false,
      tool_name: toolName,
      file_basename: basename
    });
    console.log(JSON.stringify({ status: 'pass', fail_open: true, additional_context: '', checked: false }));
    return 0;
  }

  let resolvedPath = bridgeTool ? bridgeTarget : filePath;
  if (resolvedPath.startsWith('~')) {
    const homedir = process.env.HOME || '';
    resolvedPath = path.join(homedir, resolvedPath.slice(1));
  }
  if (!path.isAbsolute(resolvedPath)) {
    resolvedPath = path.resolve(WORKSPACE_ROOT, resolvedPath);
  }

  if (!fs.existsSync(resolvedPath) || !fs.statSync(resolvedPath).isFile()) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      additional_context: 'State-watcher could not read the edited workflow-state file from disk.',
      checked: false,
      file_path: resolvedPath,
      bridge_tool: bridgeTool,
    }));
    return 0;
  }

  let document: any;
  try {
    document = JSON.parse(fs.readFileSync(resolvedPath, 'utf-8'));
  } catch (exc: any) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      additional_context: `State-watcher could not parse the workflow-state JSON: ${exc?.message || exc}.`,
      checked: false,
      file_path: resolvedPath,
      bridge_tool: bridgeTool,
    }));
    return 0;
  }

  let schema: any = {};
  if (fs.existsSync(SCHEMA_PATH) && fs.statSync(SCHEMA_PATH).isFile()) {
    try {
      schema = JSON.parse(fs.readFileSync(SCHEMA_PATH, 'utf-8'));
    } catch {
      schema = {};
    }
  }

  const errors = validateAgainstSchema(document, schema);
  let status = 'pass';
  let additionalContext = '';

  if (errors.length > 0) {
    additionalContext = (
      'State-watcher detected schema issues in the edited workflow-state document: ' +
      errors.slice(0, 8).join('; ') +
      '. Use scripts/validate-workflow-state.py for the full validator output.'
    );
    status = 'warning';
  } else {
    additionalContext = 'State-watcher confirmed the edited workflow-state document still matches .cursor/state/workflow-state.schema.json.';
    status = 'pass';
  }

  trace({
    hook: 'state-watcher',
    event: 'postToolUse',
    status,
    checked: true,
    tool_name: toolName,
    bridge_tool: bridgeTool,
    file_basename: basename,
    error_count: errors.length,
    first_errors: errors.slice(0, 4),
  });

  console.log(JSON.stringify({
    status,
    fail_open: true,
    checked: true,
    additional_context: additionalContext,
    file_path: resolvedPath,
    bridge_tool: bridgeTool,
    errors: errors.slice(0, 16),
  }).replace(/":/g, '": '));
  return 0;
}

process.exit(main());
